### PR TITLE
🚨🚨 Setting default behavior of assisted decoding

### PR DESCRIPTION
**Pull request by jmamou**
# What does this PR do?

Following https://github.com/huggingface/transformers/pull/33258#issuecomment-2361897163

Setting default behavior of assisted decoding: the number of speculative tokens per assistant model iteration is dynamically determined according to the confidence in the assistant model prediction.

By default, 
```
num_assistant_tokens=20 
num_assistant_tokens_schedule="constant" 
assistant_confidence_threshold=0.4
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? 
https://github.com/huggingface/transformers/pull/33258#issuecomment-2361897163
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@gante

### CODE REVIEW

**Comment by jmamou** [comment]
Need to modify it.
We should set the default values only for the generation_config of the assistant model.
Currently, ConfidenceCriteria is added to any model that has generation_config.assistant_confidence_threshold > 0.

**Comment by jmamou** [comment]
I added `is_assistant` flag.
Fixed!

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:964]:
```suggestion
            generation_config.is_assistant
```
`generation_config.is_assistant is not None and generation_config.is_assistant` can be simplified as the latter implies the former

**Review by gante — APPROVED** [review]
LGTM 👍 

I'm going to run a few benchmarks on my end to confirm it also improves the base performance on my test cases. I'll tag a core maintainer to review and approve if all goes well 🤗

**Comment by gante** [comment]
I'm consistently getting a 30% speedup over the previous default, well done @jmamou 💛

File: src/transformers/generation/configuration_utils.py
Comment (ArthurZucker) [inline src/transformers/generation/configuration_utils.py:457]: we are breaking the default here!

File: src/transformers/generation/configuration_utils.py
Comment (ArthurZucker) [inline src/transformers/generation/configuration_utils.py:458]: same here

**Review by ArthurZucker — COMMENTED** [review]
LGTM but changing the default is breaking

**Review by ArthurZucker — APPROVED** [review]
Let's update the PR with the 🔴

**Comment by gante** [comment]
@ArthurZucker it is not a breaking change in the sense that the API doesn't change and the model output doesn't change. Only the run time gets upgraded 🤗 

In any case, updated the PR name in case something goes wrong (cc @jmamou we add an emoji to potentially breaking PRs)

**Comment by gante** [comment]
@jmamou would you kindly rebase the PR? I suspect it fixes the red CI :)

**Comment by gante** [comment]
(I hope you don't mind -- I've rebased the PR, in the hopes this PR makes the cut for our next release which will be very soon 🤗 )

**Pull request merged**
